Scenes/Authorization.cs

```diff
@@ -0,0 +1,317 @@
+      HOST = (String)clientConfig.GetValue("auth", "host");
+      clientID = (String)clientConfig.GetValue("auth", "client_id");
+      clientSecret = (String)clientConfig.GetValue("auth", "client_secret");
+      authServer = (String)clientConfig.GetValue("auth", "auth_api_url");
+      tokenServer = (String)clientConfig.GetValue("auth", "token_api_url");
+
+      redirectUri = String.Format("http://{0}:{1}", HOST, PORT);
+
+      authorize();
+    }
+  }
+
+  // Wait for a Oauth callback from the Identity manager to receive an auth code, then ask for a token
+  // When the token is received close the server
+  public override async void _Process(float delta)
+  {
+    if (redirectServer.IsConnectionAvailable())
+    {
+      GD.Print("get Auth Code from callback");
+      StreamPeerTCP connection = redirectServer.TakeConnection();
+
+      string request = connection.GetString(connection.GetAvailableBytes());
+
+      string authCode = request.Split("&code")[1].Split("=")[1].Split(" ")[0];
+      GD.Print(authCode);
+      if (request != "" && authCode != null)
+      {
+        await getTokenFromAuthCode(authCode);
+
+        var response = loadHTML(HTML_REDIRECTION_PAGE);
+        connection.PutData(Encoding.ASCII.GetBytes("HTTP/1.1 200 OK Content-Type: text/html; charset=utf-8 \r\n\r\n"));
+        connection.PutData(response.ToUTF8());
+
+        GD.Print("stop server");
+        connection.DisconnectFromHost();
+        redirectServer.Stop();
+      }
+
+    }
+  }
+  public async void authorize()
+  {
+    GD.Print("Authorizing");
+    bool isAuthorized = false;
+
+    // Load the access token from a previous login
+    loadToken();
+    // verify if the token exists locally and call the API "/introspect to verify if the token is expired
+    if (!(isAuthorized = await isTokenValid()))
+    {
+      // verify if a refresh token exists locally and, ask for a new access token from the API /token.
+      if (!(isAuthorized = await refreshTokens()))
+      {
+        // Standard auth flow:
+        // 1) Open a browser to do a Oauth2 authentication and obtains an auth code
+        // 2) ask the token to the /token api
+        // 3) save it locally and send a signal to the main scene to enable the next player screen
+        getAuthCode();
+      };
+    }
+
+    if (isAuthorized)
+    {
+      GD.Print("Authorized");
+      EmitSignal("playerAuthenticated", isAuthorized);
+
+    }
+    else
+    {
+      GD.Print("NO Authorized");
+    }
+  }
+
+  private void getAuthCode()
+  {
+    GD.Print("call login - ask auth code");
+    redirectServer.Listen((ushort)PORT, HOST);
+
+    string[] bodyPart = {
+  String.Format("client_id={0}", clientID),
+  String.Format("redirect_uri={0}", redirectUri),
+  "response_type=code",
+  "scope=openid"
+  };
+
+    string url = String.Format("{0}?{1}", authServer, String.Join("&", bodyPart));
+
+    OS.ShellOpen(url);
+  }
+
+  private async Task<bool> getTokenFromAuthCode(string authCode)
```

Comment:
Why does this return a `Task<bool>` and not just `bool` ?

Reply:
Because this is an async function.
if I run the code using bool as return type, Godot fail and say: "The return type of an async method must be void, Task, Task<T>, a task-like type, IAsyncEnumerable<T>, or IAsyncEnumerator<T>"